app/src/main/java/protect/card_locker/importexport/CatimaImporter.java

```diff
@@ -381,28 +496,61 @@ private void importLoyaltyCard(SQLiteDatabase database, CSVRecord record)
+        String name = CSVHelpers.extractString(DBHelper.LoyaltyCardDbGroups.ID, record, null);
 
-        if (id == null) {
-            throw new FormatException("Group has no ID: " + record);
+        if (name == null) {
+            throw new FormatException("Group has no name: " + record);
         }
 
-        DBHelper.insertGroup(database, id);
+        DBHelper.insertGroup(database, name);
+    }
+
+    /**
+     * Import groups from the V3 scheme (database v17) into the database using the given session.
+     */
+    private void importGroupsV3(SQLiteDatabase database, List<CSVRecord> records) throws FormatException {
+        TreeMap<Integer, Group> sortedGroups = new TreeMap<Integer, Group>();
+        for (CSVRecord record : records) {
+            int id = CSVHelpers.extractInt(DBHelper.LoyaltyCardDbGroups.ID, record);
+            String name = CSVHelpers.extractString(DBHelper.LoyaltyCardDbGroups.NAME, record, null);
+            int order = CSVHelpers.extractInt(DBHelper.LoyaltyCardDbGroups.ORDER, record);
+
+            if (name == null) {
+                throw new FormatException("Group has no name: " + record);
+            }
+
+            DBHelper.insertGroup(database, id, name);
+            sortedGroups.put(order, DBHelper.getGroup(database, id));
+        }
+        List<Group> sortedGroupList = new ArrayList<>(sortedGroups.values());
+        DBHelper.reorderGroups(database, sortedGroupList);
     }
 
     /**
      * Import a single card to group mapping into the database using the given
      * session.
      */
-    private void importCardGroupMapping(SQLiteDatabase database, CSVRecord record) throws FormatException {
+    private void importCardGroupMappingV2(SQLiteDatabase database, CSVRecord record) throws FormatException {
         int cardId = CSVHelpers.extractInt(DBHelper.LoyaltyCardDbIdsGroups.cardID, record);
-        String groupId = CSVHelpers.extractString(DBHelper.LoyaltyCardDbIdsGroups.groupID, record, null);
+        String name = CSVHelpers.extractString(DBHelper.LoyaltyCardDbIdsGroups.groupID, record, null);
 
-        if (groupId == null) {
-            throw new FormatException("Group has no ID: " + record);
+        if (name == null) {
+            throw new FormatException("Group has no name: " + record);
         }
 
+        List<Group> cardGroups = DBHelper.getLoyaltyCardGroups(database, cardId);
+        cardGroups.add(DBHelper.getGroupByName(database, name));
+        DBHelper.setLoyaltyCardGroups(database, cardId, cardGroups);
+    }
+
+    /**
+     * Import a single card to group mapping from V3 scheme (database v17) into the database using the given
+     * session.
+     */
+    private void importCardGroupMappingV3(SQLiteDatabase database, CSVRecord record) throws FormatException {
+        int cardId = CSVHelpers.extractInt(DBHelper.LoyaltyCardDbIdsGroups.cardID, record);
+        int groupId = CSVHelpers.extractInt(DBHelper.LoyaltyCardDbIdsGroups.groupID, record);
```

Comment:
If we reuse an existing group ID in `importGroupsV3()`, we need to keep track of it to reuse it here as well.